icicle_v3/backend/cpu/include/cpu_ntt.h

```diff
@@ -0,0 +1,331 @@
+  eIcicleError coset_mul(
+    int logn,
+    int domain_max_size,
+    E* elements,
+    int batch_size,
+    const S* twiddles = nullptr,
+    int stride = 0,
+    const S* arbitrary_coset = nullptr,
+    bool bit_rev = false,
+    NTTDir dir = NTTDir::kForward,
+    bool columns_batch = false)
+  {
+    int size = 1 << logn;
+    int idx;
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_elements = elements + batch * size;
+      if (arbitrary_coset != nullptr) {
+        for (int i = 1; i < size; ++i) {
+          idx = columns_batch ? batch : i;
+          idx = bit_rev ? bit_reverse(idx, logn) : idx;
+          current_elements[i] = current_elements[i] * arbitrary_coset[idx];
+        }
+      } else if (stride != 0) {
+        for (int i = 1; i < size; ++i) {
+          idx = bit_rev ? stride * (bit_reverse(i, logn)) : stride * i;
+          idx = dir == NTTDir::kForward ? idx : domain_max_size - idx;
+          current_elements[i] = current_elements[i] * twiddles[idx];
+        }
+      }
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  eIcicleError cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output)
+  {
+    if (size & (size - 1)) { return eIcicleError::INVALID_ARGUMENT; }
+
+    // Copy input to "temp_elements" instead of pointing temp_elements to input to ensure freeing temp_elements does not
+    // free the input, preventing a potential double-free error.
+    // TODO [SHANIE]: Later, remove temp_elements and perform all calculations in-place
+    // (implement NTT for the case where columns_batch=true, in-place).
+
+    const uint64_t total_size = size * config.batch_size;
+    auto temp_elements = std::make_unique<E[]>(total_size);
+    if (config.columns_batch) {
+      transpose(input, temp_elements.get(), size, config.batch_size);
+    } else {
+      std::copy(input, input + total_size, temp_elements.get());
+    }
+    const int logn = int(log2(size));
+    const S* twiddles = CpuNttDomain<S>::s_ntt_domain.get_twiddles();
+    const int domain_max_size = CpuNttDomain<S>::s_ntt_domain.get_max_size();
+    S* arbitrary_coset = nullptr;
+    int coset_stride = 0;
+
+    if (config.coset_gen != S::one()) {
+      for (int i = 1; i <= domain_max_size; i++) {
+        if (twiddles[i] == config.coset_gen) {
+          coset_stride = i;
+          break;
+        }
+      }
+      if (coset_stride == 0) {
+        auto temp_cosets = std::make_unique<S[]>(domain_max_size + 1);
+        S coset = S::one();
+        temp_cosets[0] = coset;
+        S coset_gen = dir == NTTDir::kForward ? config.coset_gen : S::inverse(config.coset_gen);
+        for (int i = 1; i <= domain_max_size; i++) {
+          coset = coset * coset_gen;
+          temp_cosets[i] = coset;
+        }
+        arbitrary_coset = temp_cosets.get();
+        temp_cosets.release();
```

Comment:
release?
Why don't you allocate the arbitrary coset directly? as a unique_ptr I mean